docs/enhancements/agent-upgrade.md

```diff
@@ -0,0 +1,189 @@
+hosts with a version of the service, then downgrade the service to a version
+that requires an older version of the agent, and finally install the cluster
+without having to reboot the hosts.
+
+### Implementation Details/Notes/Constraints
+
+A new `compatible-agent` validation will be added to the service to check if the
+agent is compatible.
+
+The logic for that validation will be a simple string comparison of the value
+of the `AGENT_DOCKER_IMAGE` environment variable with the value of the
+`discovery_agent_version` stored in the `hosts` table.
+
+To make this comparison useful it is important that the `AGENT_DOCKER_IMAGE`
+environment variable passed to the service doesn't use the `:latest` tag, as
+that renders the comparison useless. In the _SaaS_ stage environment that
+variable is already using a tag, so that shouldn't be a problem:
+
+```
+$ oc get deployment assisted-service -n assisted-installer-stage -o json | \
+jq '.spec.template.spec.containers[].env[] | select(.name == "AGENT_DOCKER_IMAGE")'
+{
+  "name": "AGENT_DOCKER_IMAGE",
+  "value": "registry-proxy.engineering.redhat.com/rh-osbs/openshift4-assisted-installer-agent-rhel8:v1.0.0-141"
+}
+```
+
+In the integration and development environments we are using the `:latest` tag,
+which effectively disables this feature.
+
+The description of the `compatible-agent` validation will be the following:
+
+> The installation cannot start just yet because this host's agent is in the
+> process of being upgraded to a newer version, please wait, this could take
+> a few minutes.
+
+A new `upgrade-agent` step will be added to the service and to the agent.
+
+When the service receives a request for next steps from a host it will check if
+that host is in one of the `preparing-for-installation`, `installing` or
+`installing-in-progress` states. In it is then the service will return the next
+steps as usual. If it isn't then the service will retrieve from the database the
+result of the `compatible-agent` validation. If the validation passed then the
+service will return the next steps as usual, otherwise it will replace all the
+usual steps the `upgrade-agent` step.
+
+When the agent receives the `agent-upgrade` step it will run the equivalent of
+`systemctl restart agent`. That will restart the agent service, which will
+perform again the initial host registration process, get the new agent image
+name and start the agent container using it.
+
+### Risks and Mitigations
+
+The main risk is introducing interference with the state machine that controls
+the hosts. To reduce the chances of that the upgrade will be performed only when
+the host is not in the process of installing, i.e. not in one of the
+`preparing-for-installation`, `installing` or `installing-in-progress` states.
+
+## Design Details
+
+None, see the implementation details above.
+
+### Open Questions
+
+- Do the _on premise_ installations use values of `AGENT_DOCKER_IMAGE` with the
```

Comment:
I think there's nothing we can do about this other than warn users in the documentation about "don't use moving image tags". There's no way we could possibly know whether a tag is moving or not so no way for us to completely warn / block users doing this.

Reply:
The user won't typically  change the definitions for the agent, so I think we can expect sha-based image on-prem.
We can add a warning saying we might not successfully validate the agent version comparability in case the agent image is tag-based.

Reply:
@eranco74 where will this warning be? In the log? In the UI?

Reply:
Done.